Glasgow | Prati Amalden | Module Legacy Code | New feature: Rebloom

PratiAmalden opened the pull request.

PratiAmalden committed:
Add columns for rebloom

PratiAmalden committed:
feat(blooms): add rebloom creation, counts, and return original username for reblooms

PratiAmalden committed:
feat(api): add rebloom endpoint

PratiAmalden committed:
feat(ui): display rebloom info and counts in timeline

PratiAmalden committed:
feat(api): add postRebloom client request

PratiAmalden committed:
feat(ui) add rebloom elements to bloom template

PratiAmalden committed:
style(ui) add minimal styling for rebloom info

illicitonion commented on db/schema.sql:
```diff
@@ -13,6 +13,11 @@ CREATE TABLE blooms (
+ALTER TABLE blooms
+ADD COLUMN rebloom_from BIGINT REFERENCES blooms(id),
+ADD COLUMN rebloom_by INT REFERENCES users(id),
+ADD COLUMN rebloom_count INT DEFAULT 0;
```
Ideally you wouldn't need to track rebloom counts in the database - there is a race condition here where values can get out of date (e.g. if you create a rebloom but haven't updated this count yet).

Instead, use SQL relationships to compute this - in a query, you should be able to compute the count of blooms which have this ID as an rebloom_from

Your Python and JS code for querying a Bloom shouldn't have to change, but it removes the need for you to update the original bloom from the python side.

illicitonion commented on front-end/components/bloom.mjs:
```diff
@@ -84,4 +117,26 @@ function _formatTimestamp(timestamp) {
+    await apiService.postRebloom(bloom);
+
+  } catch (error) {
+    console.error("Rebloom failed:", error);
```
This isn't a user-facing error - how could you surface this to a user?

illicitonion reviewed (commented):
This is generally looking good, but there's one big thing about the database structure to look at.

From a UX perspective - it's unclear when a rebloom should sort in a timeline - should it be at time of original post or time of original bloom?

Also on a bloom when it says "Rebloomed 1 times", it may be useful to be able to click to display which users rebloomed or similar.

But generally, this is looking really good, well done!

PratiAmalden committed:
feat(blooms): prevent same user reblooming twice and track rebloom count via SQL relationships

- Updated `get_blooms_for_user` and `get_bloom` to calculate `rebloom_count`
- Uses SQL joins and `COUNT(*)` on `blooms` table with `rebloom_from` relationships to determine how many times a bloom has been rebloomed.
- Updated `add_rebloom` to raise `ValueError` when a user tries to rebloom the same original bloom twice.

PratiAmalden committed:
feat(blooms): add helper to fetch rebloomer usernames

PratiAmalden committed:
feat(blooms): add endpoint to fetch rebloomers of a bloom

PratiAmalden committed:
refactor(bloom-ui): split rebloom rendering & count logic

- Added click handler to show the list of rebloomers.
- Added proper error handling via `handleErrorDialog`.

PratiAmalden committed:
feat(api): add getRebloomers helper

PratiAmalden committed:
db(blooms): add unique constraint for reblooms

- Added UNIQUE constraint (rebloom_from, rebloom_by) to prevent duplicate reblooms.

PratiAmalden commented:
I’ve removed rebloom_count and used SQL relationships, for the timeline reblooms appear based on the original bloom time.
The “Rebloomed X times” display is now clickable and for the user-facing error I used handleErrorDialog
